### PR TITLE
Improve performance of `limitedParallelismRoundTripper`

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR improves the performance of `limitedParallelismRoundTripper` by reducing the number of goroutines spawned.

This change is based on https://github.com/grafana/loki/pull/9923.

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [n/a] Tests updated
- [n/a] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by charleskorn** [comment]
Note to reviewers: I'm not sure a changelog entry is necessary for this, but happy to add one if you disagree.

**Review by pracucci — APPROVED** [review]
LGTM! Do we have any benchmark to run? Based on the result we could then decide whether it's worth to add a CHANGELOG entry or not.

**Comment by charleskorn** [comment]
> Do we have any benchmark to run?

There's no existing benchmark, but I'll add one.

**Comment by charleskorn** [comment]
tl;dr: latency overhead is improved by up to 20% in some cases

The benchmark runs a varying number of requests and sub-requests with sub-requests that take 20ms with max parallelism set to 10, so in a perfect scenario with no overhead, we'd expect:

* requests with ten or fewer sub-requests to finish in 20ms
* requests with 20 sub-requests to finish in 40ms
* requests with 50 sub-requests to finish in 100ms
* requests with 100 sub-requests to finish in 200ms 

I've then taken the benchmark results and subtracted the theoretical time to arrive at the overhead before and after this change:

```
Name                                                                                                                                 Before  After  Theoretical time  Overhead before  Overhead after  Change
LimitedParallelismRoundTripper/1_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      21.26   21.16                20             1.26            1.16   -7.94%
LimitedParallelismRoundTripper/1_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      21.27   21.15                20             1.27            1.15   -9.45%
LimitedParallelismRoundTripper/1_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      21.29   21.17                20             1.29            1.17   -9.30%
LimitedParallelismRoundTripper/1_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     21.34   21.22                20             1.34            1.22   -8.96%
LimitedParallelismRoundTripper/1_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     42.39   42.33                40             2.39            2.33   -2.51%
LimitedParallelismRoundTripper/1_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     105.6   105.6               100              5.6             5.6   -0.00%
LimitedParallelismRoundTripper/1_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    210.8   210.6               200             10.8            10.6   -1.85%
LimitedParallelismRoundTripper/10_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     21.53   21.31                20             1.53            1.31  -14.38%
LimitedParallelismRoundTripper/10_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     21.68   21.36                20             1.68            1.36  -19.05%
LimitedParallelismRoundTripper/10_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     21.98   21.54                20             1.98            1.54  -22.22%
LimitedParallelismRoundTripper/10_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    22.18   21.87                20             2.18            1.87  -14.22%
LimitedParallelismRoundTripper/10_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    43.56   43.27                40             3.56            3.27   -8.15%
LimitedParallelismRoundTripper/10_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    107.3   106.8               100              7.3             6.8   -6.85%
LimitedParallelismRoundTripper/10_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10   213.6   211.8               200             13.6            11.8  -13.24%
LimitedParallelismRoundTripper/100_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    24.04   23.14                20             4.04            3.14  -22.28%
LimitedParallelismRoundTripper/100_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    24.57   23.77                20             4.57            3.77  -17.51%
LimitedParallelismRoundTripper/100_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    25.07   24.24                20             5.07            4.24  -16.37%
LimitedParallelismRoundTripper/100_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10   25.11   24.34                20             5.11            4.34  -15.07%
LimitedParallelismRoundTripper/100_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10   47.45   49.08                40             7.45            9.08  +21.88%
LimitedParallelismRoundTripper/100_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10   113.5   113.3               100             13.5            13.3   -1.48%
LimitedParallelismRoundTripper/100_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10  225.6   223.8               200             25.6            23.8   -7.03%
```

<details>
<summary>Detailed benchmark results</summary>

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/frontend/querymiddleware
                                                                                                                                    │ before.txt  │             after.txt             │
                                                                                                                                    │   sec/op    │   sec/op     vs base              │
LimitedParallelismRoundTripper/1_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10       21.26m ± 0%   21.16m ± 0%  -0.46% (p=0.009 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10       21.27m ± 0%   21.15m ± 0%  -0.55% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10       21.29m ± 1%   21.17m ± 0%  -0.56% (p=0.009 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      21.34m ± 1%   21.22m ± 0%  -0.56% (p=0.009 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      42.39m ± 0%   42.33m ± 1%       ~ (p=0.394 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      105.6m ± 0%   105.6m ± 0%       ~ (p=0.394 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     210.8m ± 0%   210.6m ± 1%       ~ (p=0.310 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      21.53m ± 1%   21.31m ± 1%  -1.01% (p=0.009 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      21.68m ± 1%   21.36m ± 2%  -1.47% (p=0.026 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      21.98m ± 1%   21.54m ± 0%  -2.01% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     22.18m ± 1%   21.87m ± 1%  -1.39% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     43.56m ± 1%   43.27m ± 0%  -0.66% (p=0.004 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     107.3m ± 1%   106.8m ± 1%       ~ (p=0.310 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    213.6m ± 1%   211.8m ± 1%       ~ (p=0.093 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     24.04m ± 1%   23.14m ± 7%       ~ (p=0.240 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     24.57m ± 5%   23.77m ± 4%  -3.26% (p=0.026 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     25.07m ± 2%   24.24m ± 2%  -3.32% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    25.11m ± 2%   24.34m ± 6%       ~ (p=0.132 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    47.45m ± 3%   49.08m ± 1%  +3.44% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    113.5m ± 1%   113.3m ± 1%       ~ (p=0.937 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10   225.6m ± 2%   223.8m ± 1%       ~ (p=0.180 n=6)
geomean                                                                                                                               42.98m        42.54m       -1.03%

                                                                                                                                    │  before.txt   │              after.txt              │
                                                                                                                                    │     B/op      │     B/op      vs base               │
LimitedParallelismRoundTripper/1_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10       5.891Ki ± 10%   4.205Ki ± 1%  -28.61% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10       8.102Ki ±  2%   6.635Ki ± 1%  -18.11% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10       15.81Ki ±  1%   13.91Ki ± 1%  -11.96% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      28.47Ki ±  0%   25.95Ki ± 0%   -8.86% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      53.12Ki ±  0%   52.24Ki ± 1%   -1.67% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      128.4Ki ±  2%   130.8Ki ± 1%   +1.90% (p=0.015 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     252.0Ki ±  1%   259.3Ki ± 2%   +2.90% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      54.35Ki ±  0%   41.15Ki ± 0%  -24.30% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      79.88Ki ±  0%   65.26Ki ± 0%  -18.30% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      156.2Ki ±  0%   137.6Ki ± 0%  -11.93% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     283.2Ki ±  0%   258.2Ki ± 0%   -8.84% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     530.5Ki ±  0%   517.0Ki ± 0%   -2.54% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     1.248Mi ±  1%   1.267Mi ± 1%   +1.57% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    2.488Mi ±  1%   2.553Mi ± 1%   +2.60% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     552.9Ki ±  1%   411.2Ki ± 0%  -25.62% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     809.6Ki ±  0%   652.3Ki ± 0%  -19.43% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     1.540Mi ±  0%   1.343Mi ± 0%  -12.76% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    2.783Mi ±  0%   2.520Mi ± 0%   -9.45% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    5.263Mi ±  1%   5.051Mi ± 0%   -4.03% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    12.86Mi ±  0%   13.15Mi ± 1%   +2.25% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10   25.68Mi ±  0%   26.46Mi ± 0%   +3.04% (p=0.002 n=6)
geomean                                                                                                                               321.9Ki         290.6Ki        -9.72%

                                                                                                                                    │  before.txt  │             after.txt              │
                                                                                                                                    │  allocs/op   │  allocs/op   vs base               │
LimitedParallelismRoundTripper/1_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10         79.50 ± 7%    64.00 ± 0%  -19.50% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10        113.00 ± 1%    96.00 ± 0%  -15.04% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10         215.0 ± 0%    192.0 ± 0%  -10.70% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10        385.0 ± 0%    352.0 ± 0%   -8.57% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10        716.0 ± 0%    706.0 ± 0%   -1.40% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10       1.715k ± 1%   1.765k ± 0%   +2.92% (p=0.002 n=6)
LimitedParallelismRoundTripper/1_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      3.373k ± 0%   3.525k ± 0%   +4.52% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10        775.0 ± 0%    631.0 ± 0%  -18.58% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10       1116.5 ± 0%    951.0 ± 0%  -14.82% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10       2.139k ± 0%   1.912k ± 0%  -10.63% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      3.844k ± 0%   3.514k ± 0%   -8.59% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      7.160k ± 0%   7.038k ± 0%   -1.69% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      17.17k ± 0%   17.65k ± 0%   +2.79% (p=0.002 n=6)
LimitedParallelismRoundTripper/10_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     34.13k ± 1%   35.50k ± 0%   +4.00% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_1_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      7.851k ± 1%   6.307k ± 0%  -19.67% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_2_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     11.285k ± 0%   9.509k ± 0%  -15.74% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_5_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10      21.56k ± 0%   19.12k ± 0%  -11.33% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_10_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     38.64k ± 0%   35.13k ± 0%   -9.08% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_20_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     72.53k ± 0%   70.42k ± 0%   -2.91% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_50_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10     176.0k ± 0%   181.8k ± 1%   +3.30% (p=0.002 n=6)
LimitedParallelismRoundTripper/100_concurrent_requests_with_100_sub_requests_each,_max_parallelism_10,_sub_request_duration_20ms-10    350.1k ± 0%   365.5k ± 0%   +4.39% (p=0.002 n=6)
geomean                                                                                                                                4.385k        4.063k        -7.34%

```

<details>
